IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m

```diff
@@ -26,21 +26,321 @@
+{
+    NSURL *_endUrl;
+    MSIDWebUICompletionHandler _completionHandler;
+    
+    NSLock *_completionLock;
+    NSTimer *_spinnerTimer; // Used for managing the activity spinner
+}
 
+- (id)initWithStartUrl:(NSURL *)startUrl
+                endURL:(NSURL *)endUrl
+               webview:(WKWebView *)webview
+         configuration:(MSIDWebviewConfiguration *)configuration
+               context:(id<MSIDRequestContext>)context
+{
+    self = [super initWithContext:context];
+    
+    if (self)
+    {
+        self.webView = webview;
+        _startURL = startUrl;
+        _endUrl = endUrl;
+        
+        _completionLock = [[NSLock alloc] init];
+        self.complete = NO;
+        
 #if TARGET_OS_IPHONE
-@synthesize  parentViewController;
+        self.parentController = configuration.parentController;
+        self.presentationType = configuration.presentationType;
 #endif
+    }
+    
+    return self;
+}
+
+-(void)dealloc
+{
+    [self.webView setNavigationDelegate:nil];
+    self.webView = nil;
+}
 
 - (BOOL)startWithCompletionHandler:(MSIDWebUICompletionHandler)completionHandler
 {
-    return NO;
+    // If we're not on the main thread when trying to kick up the UI then
+    // dispatch over to the main thread.
+    if (![NSThread isMainThread])
+    {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self startWithCompletionHandler:completionHandler];
+        });
+        return YES;
+    }
+    
+    // Save the completion block
+    _completionHandler = [completionHandler copy];
+    
+    NSError *error = nil;
+    [self loadView:&error];
+    if (error)
+    {
+        [self endWebAuthWithError:error orURL:nil];
+        return YES;
+    }
+    
+    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:_startURL];
+#if TARGET_OS_IPHONE
+    // Currently Apple has a bug in iOS about WKWebview handling NSURLAuthenticationMethodClientCertificate.
+    // It swallows the challenge response rather than sending it to server.
+    // Therefore we work around the bug by using PKeyAuth for WPJ challenge in iOS
+    [request addValue:kMSIDPKeyAuthHeaderVersion forHTTPHeaderField:kMSIDPKeyAuthHeader];
+#endif
+    [self startRequest:request];
+    return YES;
 }
 
 - (void)cancel
 {
+    MSID_LOG_INFO(self.context, @"Cancel Web Auth...");
+    
+    // Dispatch the completion block
+    NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorUserCancel, @"The user/application has cancelled the authorization.", nil, nil, nil, self.context.correlationId, nil);
+    [self endWebAuthWithError:error orURL:nil];
+}
+
+- (BOOL)loadView:(NSError **)error
+{
+    // create and load the view if not provided
+    BOOL result = [super loadView:error];
+    
+    self.webView.navigationDelegate = self;
+    
+    return result;
+}
+
+- (BOOL)endWebAuthWithError:(NSError *)error
+                      orURL:(NSURL *)endURL
+{
+    dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
I believe the same check is done in dispatchCompletionBlock method. So, do we need this dispatch here?

Reply:
you are right. Removed :)

Reply:
I reverted back the change for it because I realized that `dissmissWebview` in  `[self dismissWebview:^{[self dispatchCompletionBlock:endURL error:error];}];` also needs to be in main thread, so the dispatch is needed here  :)